meterfs-migrate/migrate.c

```diff
@@ -0,0 +1,673 @@
+		memcpy(h, &temp, sizeof(*h));
+		ret = 1;
+	}
+
+	if (h->id.nvalid != 0) {
+		return -1;
+	}
+
+	return ret;
+}
+
+
+static int getOldFiles(FILE *part, int whichHeader, uint32_t count, fileheader_t f[])
+{
+	uint32_t i;
+
+	for (i = 0; i < count; ++i) {
+		if (fseek(part, common.partOffset + (unsigned int)whichHeader * HEADER_SECTOR_CNT * SIZE_SECTOR + ((i + 1) * HGRAIN), SEEK_SET) != 0) {
+			return -1;
+		}
+
+		if (fread(&f[i], sizeof(*f), 1, part) == 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+static uint32_t minSectorCnt(fileheader_t *f)
+{
+	uint32_t nrecords = f->filesz / f->recordsz;
+	uint32_t bytes = nrecords * (f->recordsz + sizeof(entry_t));
+
+	return ((bytes + (SIZE_SECTOR - 1)) / SIZE_SECTOR) + 1;
+}
+
+
+static int copySector(FILE *part, uint32_t dest, uint32_t src)
+{
+	if (fseek(part, common.partOffset + (src * SIZE_SECTOR), SEEK_SET) != 0) {
+		return -1;
+	}
+
+	if (fread(common.buff, SIZE_SECTOR, 1, part) == 0) {
+		return -1;
+	}
+
+	if (fseek(part, common.partOffset + (dest * SIZE_SECTOR), SEEK_SET) != 0) {
+		return -1;
+	}
+
+	if (fwrite_workaround(common.buff, SIZE_SECTOR, 1, part) == 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int eraseSector(FILE *part, uint32_t sector)
+{
+	memset(common.buff, 0xff, SIZE_SECTOR);
+
+	if (fseek(part, common.partOffset + (sector * SIZE_SECTOR), SEEK_SET) != 0) {
+		return -1;
+	}
+
+	if (fwrite_workaround(common.buff, SIZE_SECTOR, 1, part) == 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int moveSectors(FILE *part, size_t from, size_t nsectors, size_t diff)
+{
+	int i;
+
+	for (i = nsectors; i >= 0; --i) {
+		if (copySector(part, from + i + diff, from + i) < 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+static int realocateFile(FILE *part, int id, int total, fileheader_t f[])
+{
+	uint32_t sectorsNeeded = minSectorCnt(&f[id]);
+	uint32_t diff;
+	int i, j;
+
+	if (sectorsNeeded <= f[id].sectorcnt) {
+		/* nothing to do */
+		return 0;
+	}
+
+	diff = sectorsNeeded - f[id].sectorcnt;
+
+	/* Move all files above */
+	for (i = total - 1; i > id; --i) {
+		if (moveSectors(part, f[i].sector, f[i].sectorcnt, diff) < 0) {
+			return -1;
+		}
+
+		f[i].sector += diff;
+	}
+
+	for (i = 0; i < diff; ++i) {
+		if (eraseSector(part, f[id].sector + f[id].sectorcnt - 1 + diff) < 0) {
+			return -1;
+		}
+	}
+
+	f[id].sectorcnt += diff;
+
+	return 0;
+}
+
+
+static int updateHeader(FILE *part, header_old_t *oldheader, fileheader_t *f)
+{
+	header_t newheader;
+	uint32_t i, j, checksum = 0;
+
+	for (i = 0; i < oldheader->filecnt; ++i) {
+		checksum ^= calcChecksum(&f[i], sizeof(*f));
+	}
+
+	newheader.id.no = oldheader->id.no;
+	newheader.id.nvalid = 0;
+	memcpy(newheader.magic, magicConst, sizeof(newheader.magic));
+	newheader.filecnt = oldheader->filecnt;
+	newheader.checksum = 0;
+
+	checksum ^= calcChecksum(&newheader, sizeof(newheader));
+	newheader.checksum = checksum;
+
+	for (i = 0; i < 2; ++i) {
```

Comment:
Magic number occurs here and in the rest of the code as well.

Reply:
TBH this 2 is only somewhat magic number I see. I don't really want to define this value, it's used exactly once and it's not really magic - there are two headers, this is not configurable, other parts of code assume two headers without loops